src/main/scala/scalafix/sbt/ScalafixPlugin.scala

```diff
@@ -88,6 +96,33 @@ object ScalafixPlugin extends AutoPlugin {
     def scalafixConfigSettings(config: Configuration): Seq[Def.Setting[_]] =
       Seq(
         scalafix := scalafixInputTask(config).evaluated,
+        compile := {
+          val oldCompile = compile.value
+          val scalafixRunExplicitly = executionRoots.value.exists { root =>
+            Seq(scalafix.key, scalafixAll.key).contains(root.key)
+          }
+          val runScalafixAfterCompile =
+            scalafixOnCompile.value && !scalafixRunExplicitly
+          val currentState = state.value
+          val ref = thisProjectRef.value
+          if (runScalafixAfterCompile) {
+            // We can't just run scalafix from here, as it would introduce a cyclic dependency
+            // since semantic rules require compilation products. Instead we run scalafix within
+            // a custom state where compile has not been overriden.
```

Comment:
Where's the cycle?  `val oldCompile = compile.value` means "compile and store the result (`CompileAnalysis`) as `oldCompile`.  So scalafix should find all the compilation products it needs in the right places in the target directories.

Reply:
on master: scalafix -> [fullClasspath](https://github.com/scalacenter/sbt-scalafix/blob/v0.9.18/src/main/scala/scalafix/sbt/ScalafixPlugin.scala#L346) -> compile
introduced in this PR: compile -> scalafix

Reply:
OK, instead of redefining `compile`, try something like:

```scala
scalafix := Def.taskDyn {
  // ...
  if (runScalafixAfterCompile)
    scalafix.triggeredBy(compile.in(config))
  else
    scalafix
}.value
```

Reply:
I tried that earlier but as far as I remember, it did not give me what I wanted: a failure in `scalafix` woudn't cascade to `compile`, given the fire-and-forget semantics of `triggeredBy`.

Reply:
I see, yes, that is to be expected.

So, thinking back, why does scalafix need fullClasspath but compile doesn't?  It looks like (ignoring a bunch of noise and detail) compile is defined in terms of `dependencyClasspath` and `classDirectory`, and `products` (a dependency of `fullClasspath`) is defined as 

```scala
  def makeProducts: Initialize[Task[Seq[File]]] = Def.task {
    compile.value
    copyResources.value
    classDirectory.value :: Nil
  }
```

So I wonder if your redefinition of compile could run old `compile` and `copyResources`, and then run `scalafix`, having redefined `scalafix` as depending on `dependencyClasspath` and `classDirectory`, the last of which should have the classfiles from `compile` and the resources from `copyResources`.

Reply:
When `runScalafixAfterCompile == true` then `scalafix` would be an alias of `compile`, I guess.  Or something like that.

Reply:
> having redefined scalafix as depending on dependencyClasspath and classDirectory, the last of which should have the classfiles from compile and the resources from copyResources.

By removing that dependency, the problem is that `scalafix` could run against stale class/semanticDB files, right?

Reply:
I guess there is something to do with `scalafixRunExplicitly`, which can control in which direction the dependency is so it's never cyclic for a given task execution, I'll give it a try.

Reply:
Another way to see it is: given scalafix will always depend on compilation, then "runScalafixAfterCompile" roughly means "redefine `compile` as `scalafix`".

Reply:
I think I came up with a reasonable implementation  It is a bit more tricky than I anticipated because the real `scalafix` can run any rule given that it's a input key, while the `scalafixOnCompile` can only trigger a no-arg `scalafix` which uses `.scalafix.conf` to find rules to run. That's why I changed a bit the behavior (see last paragraph of first comment).

Is that what you had in mind?

Reply:
Yeah, looks good (only fast-scanned, this late at night 😄).

Nice to know it actually worked!  (famous last words...)